[CI] AsyncShardFetchTests testTwoNodesRemoveOne failing

**Build scan:**
https://gradle-enterprise.elastic.co/s/tzalctm47olle/tests/:server:test/org.elasticsearch.gateway.AsyncShardFetchTests/testTwoNodesRemoveOne

**Reproduction line:**
```
./gradlew ':server:test' --tests "org.elasticsearch.gateway.AsyncShardFetchTests.testTwoNodesRemoveOne" -Dtests.seed=9155673A51415810 -Dtests.locale=zh-TW -Dtests.timezone=Etc/GMT-6 -Druntime.java=17
```

**Applicable branches:**
main

**Reproduces locally?:**
No, however it seems to reproduce if executed in a cycle.
Additionally each test run lasts 10 seconds seemingly due to this line: 
https://github.com/elastic/elasticsearch/blob/8cd71706faa8c28b3bed69b9729a4bd1b6b7d346/server/src/test/java/org/elasticsearch/gateway/AsyncShardFetchTests.java#L474

**Failure history:**
https://gradle-enterprise.elastic.co/scans/tests?tests.container=org.elasticsearch.gateway.AsyncShardFetchTests&tests.test=testTwoNodesRemoveOne

**Failure excerpt:**
```
com.carrotsearch.randomizedtesting.UncaughtExceptionError: Captured an uncaught exception in thread: Thread[id=1719, name=elasticsearch[testTwoNodesRemoveOne][generic][T#3], state=RUNNABLE, group=TGRP-AsyncShardFetchTests]

  at __randomizedtesting.SeedInfo.seed([9155673A51415810:1672460D3C188CD8]:0)

  Caused by: java.lang.AssertionError: (No message provided)

    at __randomizedtesting.SeedInfo.seed([9155673A51415810]:0)
    at org.junit.Assert.fail(Assert.java:86)
    at org.junit.Assert.assertTrue(Assert.java:41)
    at org.junit.Assert.assertTrue(Assert.java:52)
    at org.elasticsearch.gateway.AsyncShardFetchTests$TestFetch$1.lambda$run$0(AsyncShardFetchTests.java:474)
    at org.elasticsearch.test.ESTestCase.assertBusy(ESTestCase.java:1154)
    at org.elasticsearch.test.ESTestCase.assertBusy(ESTestCase.java:1127)
    at org.elasticsearch.gateway.AsyncShardFetchTests$TestFetch$1.run(AsyncShardFetchTests.java:474)
    at org.elasticsearch.common.util.concurrent.ThreadContext$ContextPreservingRunnable.run(ThreadContext.java:891)
    at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
    at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
    at java.lang.Thread.run(Thread.java:833)

```

--- a/server/src/test/java/org/elasticsearch/gateway/AsyncShardFetchTests.java
+++ b/server/src/test/java/org/elasticsearch/gateway/AsyncShardFetchTests.java
@@ -27,6 +27,7 @@ import java.util.Collections;
 import java.util.Map;
 import java.util.concurrent.ConcurrentHashMap;
 import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.TimeUnit;
 import java.util.concurrent.atomic.AtomicInteger;
 
 import static java.util.Collections.emptySet;
@@ -335,7 +336,6 @@ public class AsyncShardFetchTests extends ESTestCase {
         assertThat(fetchData.getData().get(node1), sameInstance(response1_2));
     }
 
-    @AwaitsFix(bugUrl = "https://github.com/elastic/elasticsearch/issues/93729")
     public void testTwoNodesRemoveOne() throws Exception {
         DiscoveryNodes nodes = DiscoveryNodes.builder().add(node1).add(node2).build();
         test.addSimulation(node1.getId(), response1);
@@ -357,7 +357,7 @@ public class AsyncShardFetchTests extends ESTestCase {
         test.fireSimulationAndWait(node1.getId());
         // there is still another on going request, so no data
         assertThat(test.getNumberOfInFlightFetches(), equalTo(1));
-        fetchData = test.fetchData(nodes, emptySet());
+        fetchData = test.fetchData(newNodes, emptySet());
         assertThat(fetchData.hasData(), equalTo(false));
 
         // fire the second simulation, this should allow us to get the data
@@ -450,9 +450,10 @@ public class AsyncShardFetchTests extends ESTestCase {
         }
 
         public void fireSimulationAndWait(String nodeId) throws InterruptedException {
-            simulations.get(nodeId).executeLatch.countDown();
-            simulations.get(nodeId).waitLatch.await();
-            simulations.remove(nodeId);
+            final var entry = simulations.get(nodeId);
+            entry.executeLatch.countDown();
+            assertTrue(entry.waitLatch.await(10, TimeUnit.SECONDS));
+            assertSame(entry, simulations.remove(nodeId));
         }
 
         @Override
@@ -464,34 +465,24 @@ public class AsyncShardFetchTests extends ESTestCase {
         protected void asyncFetch(DiscoveryNode[] nodes, long fetchingRound) {
             for (final DiscoveryNode node : nodes) {
                 final String nodeId = node.getId();
-                threadPool.generic().execute(new Runnable() {
-                    @Override
-                    public void run() {
-                        Entry entry = null;
-                        try {
-                            entry = simulations.get(nodeId);
-                            if (entry == null) {
-                                // we are simulating a master node switch, wait for it to not be null
-                                assertBusy(() -> assertTrue(simulations.containsKey(nodeId)));
-                            }
-                            assert entry != null;
-                            entry.executeLatch.await();
-                            if (entry.failure != null) {
-                                processAsyncFetch(
-                                    null,
-                                    Collections.singletonList(new FailedNodeException(nodeId, "unexpected", entry.failure)),
-                                    fetchingRound
-                                );
-                            } else {
-                                processAsyncFetch(Collections.singletonList(entry.response), null, fetchingRound);
-                            }
-                        } catch (Exception e) {
-                            logger.error("unexpected failure", e);
-                        } finally {
-                            if (entry != null) {
-                                entry.waitLatch.countDown();
-                            }
+                final Entry entry = simulations.get(nodeId);
+                assertNotNull(node.descriptionWithoutAttributes(), entry);
+                threadPool.generic().execute(() -> {
+                    try {
+                        assertTrue(entry.executeLatch.await(10, TimeUnit.SECONDS));
+                        if (entry.failure != null) {
+                            processAsyncFetch(
+                                null,
+                                Collections.singletonList(new FailedNodeException(nodeId, "unexpected", entry.failure)),
+                                fetchingRound
+                            );
+                        } else {
+                            processAsyncFetch(Collections.singletonList(entry.response), null, fetchingRound);
                         }
+                    } catch (Exception e) {
+                        throw new AssertionError("unexpected failure", e);
+                    } finally {
+                        entry.waitLatch.countDown();
                     }
                 });
             }

